// @azure/monitor-ingestion/dist/commonjs/models/models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AggregateLogsUploadError = exports.AggregateLogsUploadErrorName = exports.KnownMonitorAudience = exports.KnownVersions = void 0;
exports.isAggregateLogsUploadError = isAggregateLogsUploadError;
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
const core_util_1 = require("@azure/core-util");
/** Service API versions */
var KnownVersions;
(function (KnownVersions) {
    /** The 2023-01-01 API version. */
    KnownVersions["V20230101"] = "2023-01-01";
})(KnownVersions || (exports.KnownVersions = KnownVersions = {}));
/**
 * Known values for Monitor Audience
 */
var KnownMonitorAudience;
(function (KnownMonitorAudience) {
    /**
     * Audience for Azure China
     */
    KnownMonitorAudience["AzureChina"] = "https://monitor.azure.cn";
    /**
     * Audience for Azure Government
     */
    KnownMonitorAudience["AzureGovernment"] = "https://monitor.azure.us";
    /**
     * Audience for Azure Public
     */
    KnownMonitorAudience["AzurePublicCloud"] = "https://monitor.azure.com";
})(KnownMonitorAudience || (exports.KnownMonitorAudience = KnownMonitorAudience = {}));
/**
 * Aggregate Upload Logs Error Name
 */
exports.AggregateLogsUploadErrorName = "AggregateLogsUploadError";
/**
 * Aggregate Error type for upload function
 */
class AggregateLogsUploadError extends Error {
    /**
     *
     * @param errors - list of {@link LogsUploadFailure}
     * @param errorMessage - error message
     */
    constructor(errors, errorMessage) {
        super(`${errorMessage}\n}`);
        this.errors = errors;
        this.name = exports.AggregateLogsUploadErrorName;
    }
}
exports.AggregateLogsUploadError = AggregateLogsUploadError;
/**
 * Typeguard for AggregateUploadLogsError
 * @param e - Something caught by a catch clause.
 */
function isAggregateLogsUploadError(e) {
    if (e instanceof AggregateLogsUploadError) {
        return true;
    }
    return (0, core_util_1.isError)(e) && e.name === exports.AggregateLogsUploadErrorName;
}
